File: problems/216_combination_sum_iii/solution.py
from typing import List


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        def backtrack(start, path, remaining):
            if len(path) == k:
                if remaining == 0:
                    res.append(path[:])
                return
            for i in range(start, 10):
                if i > remaining:
                    break
                path.append(i)
                backtrack(i + 1, path, remaining - i)
                path.pop()
        backtrack(1, [], n)
        return res

File: problems/216_combination_sum_iii/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(res, new ArrayList<>(), k, n, 1);
        return res;
    }

    private void backtrack(List<List<Integer>> res, List<Integer> path, int k, int remain, int start) {
        if (path.size() == k) {
            if (remain == 0) res.add(new ArrayList<>(path));
            return;
        }
        for (int i = start; i <= 9; i++) {
            if (i > remain) break;
            path.add(i);
            backtrack(res, path, k, remain - i, i + 1);
            path.remove(path.size() - 1);
        }
    }
}
